### PR TITLE
WIP: NF: Collections

### PR DESCRIPTION
**Pull request by bpoldrack**
Introducing collections of handles and dealing with metadata to be queried.

This is still work in progress. PR aims to show how it is intended. No comments on details yet, please. About concept and basic design decisions, comments are nevertheless welcome, of course.

Notes:
- interesting files here are mainly collection.py, collectionrepo.py, handle.py, handlerepo.py; Additions to AnnexRepo and GitRepo are less important for now.
- test cases are outdated. Will change this point as soon as they are able to show how it works.
- collection_demo.py; Aims to demonstrate how these classes are supposed to be used within datalad commands.
- Example repos: https://github.com/bpoldrack/ExampleHandleJSON, https://github.com/bpoldrack/ExampleHandleRDF, https://github.com/bpoldrack/ExampleCollection.git
- right after deciding to make the PR I noticed there are still some outdated comments in there. So, don't let yourself confuse by them. Pay more attention to the docstrings at the moment.

### CODE REVIEW

File: datalad/support/gitrepo.py [hunk 1]
```diff
@@ -157,6 +163,26 @@ def newfunc(self, files, *args, **kwargs):
...
+    files: list of str
+
+    Returns:
+    list of str
+    """
+    if not isinstance(list_, list):
+        lgr.warning(
+            "_remove_empty_items() called with non-list type: %s" % type(files))
+        return files
+    return [file_ for file_ in list_ if file_]
```
Comment (yarikoptic) [inline datalad/support/gitrepo.py]: not nit-picking but rather inspiration - alternative implementation could be
```python
return filter(bool, list_)
```
;-)

File: datalad/support/handlerepo.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+        super(HandleRepo, self).__init__(path, url, direct=direct,
+                                         runner=runner, backend=backend)
+
+        self.datalad_path = opj(self.path, '.datalad')
+        self.metadata_path = opj(self.datalad_path, 'metadata')
+        if not exists(self.datalad_path):
+            os.mkdir(self.datalad_path)
+        if not exists(self.metadata_path):
+            os.mkdir(self.metadata_path)
```
Comment (yarikoptic) [inline datalad/support/handlerepo.py]: should we get may be a 

```python
def assure_dir(*args):
    dirname = opj(*args)
    if not exists(dirname):
        os.makedirs(dirname)
    return dirname
```
?   I see it being used in few other spots ;)

File: datalad/support/gitrepo.py (hunk 1 above)
Comment (bpoldrack) [inline datalad/support/gitrepo.py]: Wasn't aware - thanks.

File: datalad/support/handlerepo.py (hunk 2 above)
Comment (bpoldrack) [inline datalad/support/handlerepo.py]: Agree

**Comment by yarikoptic** [comment]
Please merge current master

**Pull request merged**